sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -62,8 +62,9 @@ trait HashJoin {
   }
 
   protected lazy val (buildKeys, streamedKeys) = {
-    require(leftKeys.map(_.dataType) == rightKeys.map(_.dataType),
-      "Join keys from two sides should have same types")
+    require(leftKeys.map(_.dataType).zip(rightKeys.map(_.dataType)).forall {
+      case (l, r) => BinaryOperator.sameType(l, r)
```

Comment:
Do we have a test case to cover this?

Reply:
I hit a test failure before changing this. This kind of check(assert, require) can only be hitten when there is a bug.